Sources/SwiftProtobuf/Message.swift

```diff
@@ -129,6 +132,14 @@ extension Message {
     return true
   }
 
+  /// Default traverse implementation
+  public func traverse<V: Visitor>(visitor: inout V) throws {
+    for field in Self._fields {
+      try field.traverse(message: self, visitor: &visitor)
+    }
+    try unknownFields.traverse(visitor: &visitor)
```

Comment:
How would that work? Would we just not visit unknown fields at all in the implementation of traverse then? Or would this be maybe an argument to `traverse` to decide the order of traversal?

Reply:
My idea:  `traverse` wouldn't visit unknown fields at all.   The encoders would be changed to use `traverse` to visit the regular fields and request the unknown fields directly from the message.

Reply:
> traverse wouldn't visit unknown fields at all
Could that break existing implementations of `Visitor` that are expecting a call to `visitUnknown`? The compiler will not complain even if `visitUnknown` is removed from the protocol and could cause bugs to those who are expecting this to be called on their visitor implementations.

Reply:
> Could that break existing implementations of `Visitor` that are expecting a call to `visitUnknown`?

Yeah, that's a good point.  That suggests we keep `traverse` the way it is and add other methods for alternative strategies.  Perhaps `traverseBackwards`?  Or `traverseFieldsOnly`?

Reply:
Here's one rough idea of how I was thinking could `Visitor` might control the behavior:

```swift

public protocol Visitor {

  public enum UnknownMode {
    // Visit any unknowns before the fields/extensions.
    case first
    // Visit any unknowns after the fields/extensions.
    case last
    // Don't visit any unknowns.
    case skip
  }

  public static var unknownMode:UnknownMode { get }

  public enum FieldOrder {
    // Visit fields/extensions in increasing number order.
    case forward
    // Visit fields/extensions in decreasing number order.
    case reverse
  }

  public static var fieldOrder:FieldOrder { get }

  // ...
}

// Defaults
extension Visitor {
  public static var unknownMode:UnknownMode { return .last }
  public static var fieldOrder:FieldOrder { return .forward }
}

```

And then the default for `Message` `traverse()` could be something like:

```swift

  public func traverse<V: Visitor>(visitor: inout V) throws {
    let unknownMode = V.Type.unknownMode
    if unknownMode == .first {
      try unknownFields.traverse(visitor: &visitor)
    }

    switch V.Type.fieldOrder {
    case .forward:
      for field in Self._fields {
        try field.traverse(message: self, visitor: &visitor)
      }
    case .reverse:
      for field in Self._fields.reversed() {
        try field.traverse(message: self, visitor: &visitor)
      }
    }

    if unknownMode == .last {
      try unknownFields.traverse(visitor: &visitor)
    }
  }

```

Or something like that.

Reply:
This is sorta what I was originally thinking, but since the `visit*Message()` impls usually have to recurse, if they forget to forward the `options`, it would change to the `.defaultOptions` which could be a subtle bug.

I also wondered if a given _Visitor_ will always use the same options, in which case that seems like it being _properties_ of the _Visitor_ might make sense.

But I'm not tied to one vs. the other, the potential danger with having to pass through the `Options` does worry me a little.

Reply:
In practice, visitors always traverse in the same way, and call the traverse method from just one of a very few places in their source.  So I would suggest using either arguments or separate functions for the different traversal styles.  In practice, I'm pretty sure that reverse traversal will always visit unknown fields first, and forward traversal will always visit last.  The only exception is -- if I remember correctly -- JSON doesn't actually support unknown fields at all.

Reply:
If I'm remembering correctly, I think Go might have support to traverse fields within a message in random order so folks can't bake assumptions into the binary encoding (they might also space pad JSON for the same sorta reasons). Which I guess could be exposed as to the public EncodingOptions to let folks pick if they wanted the default (non deterministic) or deterministic ordering, like we have for `map<>`s. So, if that's something we ever thought we might also want to support, that might be a reason not to go two different traversal methods, but instead stick with some more of way to control it via some options.